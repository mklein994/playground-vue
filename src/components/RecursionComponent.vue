<script setup lang="ts">
import { computed } from "vue";

import { dataSet, dataSetObject, type RecursiveMap } from "@/helpers/dataSet";

// Props make this easier to test
const props = withDefaults(
  defineProps<{
    source?: string[] | [key: string, value: string][];
    split?: (word: string) => string[];
    parseStyle?: "map" | "object";
  }>(),
  {
    source: () => __PLAYGROUND_VUE_FILES_LIST__,
    split: (word: string) => word.match(/\/[^/]+/g) ?? [],
    parseStyle: "object",
  },
);

const checkHasValues = (
  value: unknown,
): value is [key: string, value: string][] => {
  return (
    Array.isArray(value)
    && value.some((x) => Array.isArray(x) && x.length === 2)
  );
};

const sourceTree = computed(() =>
  checkHasValues(props.source)
    ? props.source.reduce(
        (all, [key, value]) => dataSet(all, props.split(key), value),
        new Map() as RecursiveMap,
      )
    : props.source.reduce(
        (all, one) => dataSet(all, props.split(one), one),
        new Map() as RecursiveMap,
      ),
);

const sourceObject = computed(() =>
  checkHasValues(props.source)
    ? props.source.reduce(
        (all, [key, value]) => dataSetObject(all, props.split(key), value),
        {} as Record<string, unknown>,
      )
    : props.source.reduce(
        (all, one) => dataSetObject(all, props.split(one), one),
        {} as Record<string, unknown>,
      ),
);
</script>

<template>
  <div class="recursion">
    <pre v-if="parseStyle === 'map'">{{ sourceTree }}</pre>
    <pre v-else>{{ sourceObject }}</pre>
  </div>
</template>

<style scoped>
.recursion {
  display: grid;
  max-width: max-content;
}
</style>
