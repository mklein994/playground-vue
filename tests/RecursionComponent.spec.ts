import { flushPromises, mount } from "@vue/test-utils";
import { describe, expect, it } from "vitest";

import RecursionComponent from "@/components/RecursionComponent.vue";

describe("RecursionComponent", () => {
  // This test can't be run concurrently, since it sometimes causes the
  // snapshot to be discarded, marking it as "obsolete". I'm not sure why.
  it("renders correctly (map)", async () => {
    const wrapper = mount(RecursionComponent, {
      props: {
        source: [
          "store.produce.apple",
          "store.produce.banana",
          "store.produce.cherry",
          "person.male.dad",
          "person.male.son",
          "person.female.mom",
          "person.female.daughter",
        ],
        split: (word: string) => word.split("."),
        parseStyle: "map",
      },
    });
    await flushPromises();

    expect(wrapper.html()).toMatchSnapshot();
  });

  it("renders correctly (object)", async () => {
    const wrapper = mount(RecursionComponent, {
      props: {
        source: [
          "store.produce.apple",
          "store.produce.banana",
          "store.produce.cherry",
          "person.male.dad",
          "person.male.son",
          "person.female.mom",
          "person.female.daughter",
        ],
        split: (word: string) => word.split("."),
        parseStyle: "object",
      },
    });
    await flushPromises();

    expect(wrapper.html()).toMatchSnapshot();
  });

  it("handles an empty source", async () => {
    const wrapper = mount(RecursionComponent, {
      props: {
        source: [],
        split: (word: string) => word.split("."),
      },
    });
    await flushPromises();

    expect(
      wrapper
        .findAll("output")
        .map((x) => x.text())
        .every((x) => x === "(undefined)"),
      "not every output contains '(undefined)'",
    ).toBe(true);
  });
});
