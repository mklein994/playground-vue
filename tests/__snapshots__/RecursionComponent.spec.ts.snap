// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RecursionComponent > renders correctly (map) 1`] = `
"<div data-v-5572d211="" class="recursion"><pre data-v-5572d211="">{
  "Map(2)": {
    "store =>": {
      "Map(1)": {
        "produce =>": {
          "Map(3)": {
            "apple =>": "store.produce.apple",
            "banana =>": "store.produce.banana",
            "cherry =>": "store.produce.cherry"
          }
        }
      }
    },
    "person =>": {
      "Map(2)": {
        "male =>": {
          "Map(2)": {
            "dad =>": "person.male.dad",
            "son =>": "person.male.son"
          }
        },
        "female =>": {
          "Map(2)": {
            "mom =>": "person.female.mom",
            "daughter =>": "person.female.daughter"
          }
        }
      }
    }
  }
}</pre>
</div>"
`;

exports[`RecursionComponent > renders correctly (object) 1`] = `
"<div data-v-5572d211="" class="recursion"><pre data-v-5572d211="">{
  "store": {
    "produce": {
      "apple": "store.produce.apple",
      "banana": "store.produce.banana",
      "cherry": "store.produce.cherry"
    }
  },
  "person": {
    "male": {
      "dad": "person.male.dad",
      "son": "person.male.son"
    },
    "female": {
      "mom": "person.female.mom",
      "daughter": "person.female.daughter"
    }
  }
}</pre>
</div>"
`;
